scheduler-server/scheduling-dsl-compiler/src/main.ts

```diff
@@ -25,80 +25,69 @@ const tsConfig = JSON.parse(fs.readFileSync(new URL('../tsconfig.json', import.m
 const { options } = ts.parseJsonConfigFileContent(tsConfig, ts.sys, '');
 const compilerTarget = options.target ?? ts.ScriptTarget.ES2021
 
-const input = readline.createInterface(process.stdin);
-
-function registerInputCallback(callback: (data: string) => void) {
-  input.once('line', callback);
-}
-
-function writeOutputOneLine(data: string) {
-  process.stdout.write(data + '\n');
-}
-
-function writeSuccess(data: string) {
-  writeOutputOneLine('success');
-  writeOutputOneLine(data);
-}
-
-function writeError(data: string) {
-  writeOutputOneLine('error');
-  writeOutputOneLine(data);
-}
-
-function writePanic(data: string) {
-  writeOutputOneLine('panic');
-  writeOutputOneLine(data);
-}
-
 process.on('uncaughtException', (err) => {
   console.error('uncaughtException');
   console.error((err && err.stack) ? err.stack : err);
-  writePanic(err.stack ?? err.message);
+  process.stdout.write('panic' + '\n');
+  process.stdout.write((err.stack ?? err.message) + '\n');
   process.exit(1);
 });
 
-async function handleRequest(data: string) {
+const lineReader = readline.createInterface({
+  input: process.stdin,
+});
+lineReader.once('line', handleRequest);
+
+async function handleRequest(data: Buffer) {
   try {
-    const { goalCode, schedulerGeneratedCode, constraintsGeneratedCode } = JSON.parse(data) as { goalCode: string, schedulerGeneratedCode: string, constraintsGeneratedCode: string };
+    // Test the health of the service by responding to "ping" with "pong".
+    if (data.toString() === 'ping') {
+      process.stdout.write('pong\n');
+      lineReader.once('line', handleRequest);
+      return;
+    }
```

Comment:
Good question - this check is only between the Java process and the node subprocess (so the UI can't send a "ping" message). The check is mostly supposed to be a "fail fast" mechanism, so that if the node subprocess is misconfigured, we find out at startup time rather than waiting for the first compilation request to come in.

Now, unfortunately, if the subprocess is unresponsive, the whole application hangs, since we haven't figured out how to set a timeout.

Reply:
This is at least partly handled by the user code runner having a timeout, but the process as a whole does not.